### PR TITLE
Add unit tests for deserializing a comprehensive set of avro/connect …

### PR DESCRIPTION
**Pull request by rodesai**
…types

This patch adds unit tests for deserializing the various avro and connect types
into their compatible ksql types. It also includes fixes for a few issues found
from testing:
    - conversion of int8/16/32 to larger integer types
    - conversion of logical types into valid ksql types
    - better error message when there is a type mismatch

### Description 
_What behavior do you want to change, why, how does your patch achieve the changes?_

### Testing done 
_Describe the testing stategy. Unit and integration tests are expected for any behavior changes._

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroDeserializerTest.java [hunk 1]
```diff
@@ -201,4 +213,342 @@ public void shouldDeserializeWithMissingFields() {
 
     return kafkaAvroSerializer.serialize(topicName, avroRecord);
   }
+
+  private void shouldDeserializeTypeCorrectly(final org.apache.avro.Schema avroSchema,
+                                              final Object avroValue,
+                                              final Schema ksqlSchema) {
+    shouldDeserializeTypeCorrectly(avroSchema, avroValue, ksqlSchema, avroValue);
```
Comment (hjafarpour) [inline ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroDeserializerTest.java]: Why do we use `avroValue` as `ksqlValue`?

File: ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroDeserializerTest.java [hunk 2]
```diff
@@ -201,4 +213,342 @@ public void shouldDeserializeWithMissingFields() {
...
+
+  @Test
+  public void shouldDeserializeConnectInt16ToBigint() {
+    shouldDeserializeConnectTypeCorrectly(
+        Schema.INT16_SCHEMA,
+        (short) 16384,
+        Schema.OPTIONAL_INT64_SCHEMA,
+        16384L
+    );
+  }
```
Comment (hjafarpour) [inline ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroDeserializerTest.java]: A test to validate the correct column name in error message would be nice to have.

File: ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroDeserializerTest.java (hunk 1 above)
Comment (rodesai) [reply ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroDeserializerTest.java]: For many cases the expected value is the same as the value in the avro record.

File: ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroDeserializerTest.java (hunk 2 above)
Comment (rodesai) [reply ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroDeserializerTest.java]: This was added to ConnectDataTranslatortest

File: ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java [hunk 3]
```diff
@@ -52,60 +58,145 @@ public GenericRow toKsqlRow(final Schema connectSchema,
...
+      case INT32:
+        validateType(
+            pathStr, schema, connectSchema,
+            Schema.Type.INT32, Schema.Type.INT16, Schema.Type.INT8);
+        break;
+      case FLOAT64:
+        validateType(pathStr, schema, connectSchema, Schema.Type.FLOAT32, Schema.Type.FLOAT64);
+        break;
+      default:
+        throw new RuntimeException(
```
Comment (dguy) [inline ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java]: Should this be `UnsupportedTypeException` as in another one of your PRs?

File: ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java [hunk 4]
```diff
@@ -52,60 +58,145 @@ public GenericRow toKsqlRow(final Schema connectSchema,
...
+            connectSchema.type().getName(),
+            schema.type().getName(),
+            pathStr));
+  }
+
+  private void validateType(final String pathStr,
+                            final Schema schema,
+                            final Schema connectSchema,
+                            final Schema.Type... validTypes) {
+    Arrays.asList(validTypes).stream()
```
Comment (dguy) [inline ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java]: `Arrays.stream(..)`

File: ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java [hunk 5]
```diff
@@ -52,60 +58,145 @@ public GenericRow toKsqlRow(final Schema connectSchema,
...
+        .filter(connectSchema.type()::equals)
+        .findFirst()
+        .orElseThrow(
+            () -> createTypeMismatchException(pathStr, schema, connectSchema));
+  }
+
+  private void validateType(final String pathStr,
```
Comment (dguy) [inline ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java]: Do we need this method? It is only called from 1 place AFAICT

File: ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroDeserializerTest.java [hunk 6]
```diff
@@ -201,4 +213,342 @@ public void shouldDeserializeWithMissingFields() {
...
+        .name("field0")
+        .type(avroSchema)
+        .noDefault()
+        .endRecord();
+    final Schema ksqlRecordSchema = SchemaBuilder.struct().field("field0", ksqlSchema).build();
+
+    final GenericRecord avroRecord = new GenericData.Record(avroRecordSchema);
+    avroRecord.put("field0", avroValue);
+
+    final Map map = new HashMap();
```
Comment (dguy) [inline ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroDeserializerTest.java]: nit: `Map<String, Object>`

Comment (dguy) [reply ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroDeserializerTest.java]: Also some overlap here with the same in `getSerializedRow`. Perhaps extract a method to create the `GenericRecord` ?

File: ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java (hunk 3 above)
Comment (rodesai) [reply ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java]: UnsupportedTypeException was intended to be internal to ConnectSchemaTranslator.

File: ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java (hunk 5 above)
Comment (rodesai) [reply ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java]: I'll remove it.

**Pull request merged**